design/crypto_primitives.rs

```diff
@@ -0,0 +1,179 @@
+// modified, or distributed except according to those terms. Please review the Licences for the
+// specific language governing permissions and limitations relating to use of the SAFE Network
+// Software.
+
+//! Different MaidSafe tools need different crypto primitives. E.g. Crust only uses encryption
+//! public/private keys, Routing uses encryption and signing public/private keys, etc.
+//! Currently in some cases safe_crypto is unusable cause it doesn't expose some internal data like
+//! private keys, etc.
+//! One way to solve this and stay super flexible is to have 1-to-1 mappings with lower level
+//! crypto types from rust_sodium, miscreant, etc.
+//! This document attempts to catch safe_crypto design that would work for all our our cases.
+//!
+//! `PublicKeys` and `SecretKeys` currently contain both signing and encryption keys. In some
+//! cases, we don't need signing key at all, e.g. in Crust. So this document proposes to split
+//! `PublicKeys` and `SecretKeys` into corresponding types: `PublicEncryptKey, `SecretEncryptKey`,
+//! `PublicSignKey` and `SecretSignKey`.
+
+// whatever is defined in the underlying crypto library
+const PUBLIC_ENCRYPT_KEY_BYTES: usize = rust_sodium::crypto::box_::PUBLICKEYBYTES;
+const SECRET_ENCRYPT_KEY_BYTES: usize = rust_sodium::crypto::sign::SECRETKEYBYTES;
+const PUBLIC_SIGN_KEY_BYTES: usize = rust_sodium::crypto::sign::PUBLICKEYBYTES;
+const SECRET_SIGN_KEY_BYTES: usize = rust_sodium::crypto::sign::SECRETKEYBYTES;
+const SHARED_SECRET_KEY_BYTES: usize = XXX;
+const SIGNATURE_BYTES: usize = XXX;
+const SYMMETRIC_KEY_BYTES: usize = XXX;
+
+quick_error! {
+    /// This error is returned if encryption or decryption fail.
+    /// The encryption failure is rare and mostly connected to serialisation failures.
+    /// Decryption can fail because of invalid keys, invalid data, or deserialisation failures.
+    #[derive(Debug)]
+    pub enum Error {
+        /// Occurs when serialisation or deserialisation fails.
+        Serialisation(e: SerialisationError) {
+            description("error serialising or deserialising message")
+            display("error serialising or deserialising message: {}", e)
+            cause(e)
+            from()
+        }
+        /// Occurs when we can't decrypt a message or verify the signature.
+        DecryptVerify(_e: ()) {
+            description("error decrypting/verifying message")
+            from()
+        }
+        /// Occurs in case of an error during initialisation.
+        InitError(e: i32) {
+            description("error while initialising random generator")
+            from()
+        }
+        /// Occurs when we fail to derive encryption key from password.
+        DeriveKey {
+            description("error deriving encryption key from password")
+        }
+    }
+}
+
+/// Derives encryption key from plaintext password.
+pub fn key_from_password(password: &[u8], salt: &[u8]) -> Result<SymmetricKey, Error>;
```

Comment:
Key derivations usually will have a CPU and Memory complexity parameters associated - to thwart brute force attacks even for simple passwords. E.g. if deriving the correct key takes x CPU cycles == X secs and y amount of memory, then that makes brute forcing practically useless unless you run them in parallel over several machines, but anyway it's significantly decreased the ability to do it easily. If you are looking in scrypt or any other password hashing don't they have a way to specify those parameters ? How does it compare to `pwhash` from `rust_sodium` for instance ? 

Might be a useful feature we don't want to miss as our goal is not so much as to store (or encourage to store) the password in plaintext and protect it (via hashing) than to actually deter the mentioned attacks by making it computationally very dear to try permutations.

Reply:
I think the idea was to remove these configuration details, at least in a first place and only expose them if and when needed (based on Fraser's comment)
>  would guess that we don't need this KeyDeriveParams struct at all, at least for now. It'd be easier to just pick what we think are the best default values to initialise our chosen PBKDF and hide that inside key_from_password(). We can always add this extra flexibility to the safe_crypto API if there's a demand for it? Harder to remove it from the API later on.

Reply:
Yes that's already covered and actioned according to @Fraser999 's review below ? (the cpu and mem limit parameters have been removed outdating his comment)